src/main/java/org/folio/circulation/domain/notice/schedule/RequestScheduledNoticeHandler.java

```diff
@@ -0,0 +1,118 @@
+
+import org.folio.circulation.domain.Request;
+import org.folio.circulation.domain.RequestAndRelatedRecords;
+import org.folio.circulation.domain.RequestRepository;
+import org.folio.circulation.domain.notice.NoticeTiming;
+import org.folio.circulation.domain.notice.PatronNoticeService;
+import org.folio.circulation.domain.notice.TemplateContextUtil;
+import org.folio.circulation.support.Clients;
+import org.folio.circulation.support.Result;
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+
+import io.vertx.core.json.JsonObject;
+
+public class RequestScheduledNoticeHandler {
+
+  public static RequestScheduledNoticeHandler using(Clients clients) {
+    return new RequestScheduledNoticeHandler(
+      RequestRepository.using(clients, true),
+      PatronNoticeService.using(clients),
+      ScheduledNoticesRepository.using(clients));
+  }
+
+  private RequestRepository requestRepository;
+  private PatronNoticeService patronNoticeService;
+  private ScheduledNoticesRepository scheduledNoticesRepository;
+
+  private RequestScheduledNoticeHandler(RequestRepository requestRepository,
+                                        PatronNoticeService patronNoticeService,
+                                        ScheduledNoticesRepository scheduledNoticesRepository) {
+
+    this.requestRepository = requestRepository;
+    this.patronNoticeService = patronNoticeService;
+    this.scheduledNoticesRepository = scheduledNoticesRepository;
+  }
+
+  public CompletableFuture<Result<Collection<ScheduledNotice>>> handleNotices(Collection<ScheduledNotice> scheduledNotices) {
+    CompletableFuture<Result<ScheduledNotice>> future = completedFuture(succeeded(null));
+    for (ScheduledNotice scheduledNotice : scheduledNotices) {
+      future = future.thenCompose(r -> r.after(v -> handleRequestNotice(scheduledNotice)));
+    }
+    return future.thenApply(r -> r.map(v -> scheduledNotices));
+  }
+
+  private CompletableFuture<Result<ScheduledNotice>> handleRequestNotice(ScheduledNotice notice) {
+    return requestRepository.getById(notice.getRequestId())
+      .thenApply(r -> r.map(RequestAndRelatedRecords::new))
+      .thenCompose(r -> r.after(records -> sendNotice(records, notice)))
+      .thenCompose(r -> r.after(records -> updateNotice(records, notice)));
+  }
+
+  private CompletableFuture<Result<RequestAndRelatedRecords>> sendNotice(
+    RequestAndRelatedRecords relatedRecords, ScheduledNotice notice) {
+    Request request = relatedRecords.getRequest();
+
+    if (request.isClosed()) {
+      return completedFuture(succeeded(relatedRecords));
+    }
+
+    JsonObject requestNoticeContext = TemplateContextUtil.createRequestNoticeContext(request);
+
+    return patronNoticeService.acceptScheduledNoticeEvent(
+      notice.getConfiguration(), relatedRecords.getUserId(), requestNoticeContext)
+      .thenApply(r -> r.map(v -> relatedRecords));
+  }
+
+  private CompletableFuture<Result<ScheduledNotice>> updateNotice(
```

Comment:
What does update notice do?

Does it reschedule the notice if necessary? Are completed notices marked as sent, or deleted?

Reply:
Yes, it reschedules the notice if necessary. When notice is sent and do not need rescheduling than the notice is deleted.